File: framework/fit/python/fit_flowable/core/publisher.py
# -- encoding: utf-8 --
# Copyright (c) 2024 Huawei Technologies Co., Ltd. All Rights Reserved.
# This file is a part of the ModelEngine Project.
# Licensed under the MIT License. See License.txt in the project root for license information.
# ======================================================================================================================
"""
功 能：发布者，负责产生数据流并发布给订阅者。
"""
from abc import abstractmethod

from fit_flowable.core.subscriber import Subscriber


class Publisher:
    """
    表示发布者的抽象基类，发布者是可以拥有订户并向其发布事件的对象。

    * 子类必须实现订阅方法，以将新订阅者添加到发布者的订阅者列表中。
    """

    @abstractmethod
    def subscribe(self, subscriber: Subscriber) -> None:
        """
        向发布者订阅以启动数据发送。
        * 该方法可被多次执行，每次将为其订阅者产生一个新的 Subscription；
        * 在订阅过程中发生的异常将通过 Subscriber 的 fail 方法传递。

        :param subscriber: 已订阅的 Subscriber。
        """
        pass

File: framework/fit/java/fit-reactor/src/main/java/com/huawei/fitframework/flowable/Publisher.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 */

package com.huawei.fitframework.flowable;

/**
 * 表示发布者，负责产生数据流并发布给订阅者。
 * <p>发布者产生一系列数量不限的的元素，并按照其订阅者的要求进行发布，一个发布者可为多个订阅者提供服务，其订阅者通过调用其
 * {@link #subscribe(Subscriber)} 方法对其进行订阅。</p>
 *
 * @param <T> 表示发布的数据类型的 {@link T}。
 * @author 季聿阶
 * @since 2024-02-07
 */
public interface Publisher<T> {
    /**
     * 向发布者订阅以启动数据发送。
     * <p>该方法可被多次执行，每次将为其订阅者产生一个新的 {@link Subscription}，在订阅过程中发生的异常将通过
     * {@link Subscriber#fail(Exception)} 进行传递。</p>
     *
     * @param subscriber 表示已订阅的 {@link Subscriber}{@code <}{@link T}{@code >}。
     */
    void subscribe(Subscriber<T> subscriber);
}
